### PR TITLE
nv2a: Implement SET_MATERIAL_ALPHA.

### LINKED ISSUE #242: [Bug] Some textures not rendered properly in MagiDeathFight!
Xemu   

Cxbx reloaded

Xemu

Cxbx reloaded

### xemu Version

Version: 0.6.2
Branch: master
Commit: bbcd576
Date: Sat Oct 30 23:09:11 UTC 2021
Last-known Working: N/A

### System Information

Field | Value
-- | --
OS | Windows 10 64-bit (21H1)
CPU | AMD Ryzen 5 2600 Six-Core Processor @ 3.40GHz
Graphics Device | AMD Radeon RX 5700 XT
Graphics Driver | 4.0.148202 Core Profile Context 21.3.2 27.20.15003.5017

### LINKED ISSUE #397: Jet Set Radio Future: Map Bug
### Title Information

* JSRF: Jet Set Radio Future IG-024 [49470018]
* https://xemu.app/titles/49470018/

### Bug Description
The maps on the pause screen are filled in gray blocks in the shape of the map instead of showing elevations, graffiti, and such.

Steps to reproduce this issue:
1. Start game.
2. Load into a level.
3. Pause the game and look at the map.

JSRF map 0

### Expected Behavior
The maps should look like this...

https://giantbomb1.cbsistatic.com/uploads/original/1/15777/2143183-img_20120304_054531.jpg

### Version

* Version: 0.5.4-21-gf03015e20e

### System Information

| Field | Value |
|-------|-------|
| OS			|Windows 10 Pro| 
| CPU			| AMD FX(tm)-8350 Eight-Core Processor| 
| Graphics Device	|Radeon RX 580| 
| Graphics Driver	| Radeon 21.4.1| 

### Additional Context
This isn't a huge issue during early levels since they are pretty small or linear. But later maps can be big sprawling multi-tiered places where finding things without the map might turn into a "needle in a haystack" situation.

### LINKED ISSUE #700: nv2a 0x97 -> 0x03b4 (3D_MATERIAL_FACTOR_FRONT_A?) unhandled
### Title

https://xemu.app/titles/4d53006e
https://xemu.app/titles/4d5300d1

### Bug Description

According to the debug output, 0x97 -> 0x03b4 is unhandled.
In the [NV20 nouveau sources](https://gitlab.freedesktop.org/mesa/mesa/-/blob/184a690fca700aa9ef2ae2e58b0f4bd86562aa6d/src/mesa/drivers/dri/nouveau/nv20_3d.xml.h#L868), this instruction is `3D_MATERIAL_FACTOR_FRONT_A` (A as in "Alpha").

### Expected Behavior

The NV2A instruction 0x03b4 should be handled.

### xemu Version

0.6.2-55-gcaf84884c4

### System Information

N/A

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by abaire**
Adds handling for the `SET_MATERIAL_ALPHA` command and the `diffuse_src` in the fixed function lighting model.

The test suite is here: https://github.com/abaire/nxdk_pgraph_tests/blob/main/tests/material_alpha_tests.cpp

Fixes #397
Fixes #700 
Screenshot_20211217_232350
 
(From @Triticum0)  Fixes #242

### CODE REVIEW

**Comment by abaire** [comment]
I don't fully understand the `NV_IGRAPH_XF_*` registers in `nv2a_regs.h` so I'm not sure if there's actually a HW register that should be used for the alpha value. If you can point me at docs or how to read back the register from HW I can probably figure it out. For now I just created a standalone value and piped it through to the FF shader.

**Comment by Triticum0** [comment]
I did some testing and this pr also fixes #242. Thank abaire. Once #331 code is tested to be correct this game should be fully playable on Xemu. Great job!!!

**Comment by Triticum0** [comment]
This could also fix #598,#584 and  #547 but have tested yet to confirm. Will get back to you in the coming days

**Comment by abaire** [comment]
> This could also fix #598,#584 and #547 but have tested yet to confirm. Will get back to you in the coming days

Awesome, fingers crossed that this fixes a bunch of glitches and thank you very much for testing, as always!

**Comment by Shoegzer** [comment]
Fantastic find @abaire!  Out of curiosity - the fixed screenshot shows the progress indicator (lower-left on screen) with a black background rather than light blue as it should be.  Is that part of a separate issue?

**Comment by abaire** [comment]
> Fantastic find @abaire! Out of curiosity - the fixed screenshot shows the progress indicator (lower-left on screen) with a black background rather than light blue as it should be. Is that part of a separate issue?

@Shoegzer I believe it is separate. As far as I can see from the actual HW screens the light blue doesn't have any transparency. The fact that the icons also do not display makes me think there's something interesting about the way that's rendered.

It's probably worth filing a bug if you feel like it, though @Triticum0 might know if there's already a bug tracking similar black-out issues with UI elements (I'm inclined to guess this happens in more than one game).

**Comment by Shoegzer** [comment]
Thanks, and that makes sense even if unfortunate.  Let's see what @Triticum0 comes up with but I'm happy to file an issue for this.

I tested JSRF against this PR myself and found the map issue is indeed fixed.  I tested a few other issues too that I thought had some possibility of being fixed (transparent floor near beginning of Halo 2 and menu elements in Star Wars: Knights of the Old Republic) but no change so I assume separate issues.  Still, this is nice progress.

**Comment by Triticum0** [comment]
It only fixes the #598 where the colour of the characters energy gauge is missing. The other issues I mentioned aren't fixed. A bit off-topic but looking at the code, I think this issue from [#281](https://github.com/xqemu/xqemu/issues/281) may be useful in making the implementation accurate implementation as JayFoxRox never finished the work on it. It's mostly a hunch as I don't have any knowledge on the Xbox inter-workings, But either way, [jayfoxrox](https://github.com/JayFoxRox/xqemu) and [xqemu](https://github.com/xqemu/xqemu) repository has a lot of useful information that was never transferred to Xemu as issues/pr when it was forked. So you might find something useful Or ask matt as he is most knowledgeable with the code base.

**Comment by abaire** [comment]
After extending my test cases a bit more, the behavior this introduces is not exactly right. I think this is tied to the fact that the current shader doesn't respect the diffuse source selection - if it's set to material, then this PR is correct, but if it's set to diffuse or specular the material alpha is erroneously applied. I think we should fix that before merging to avoid introducing regressions.

**Comment by abaire** [comment]
Updated to respect the `diffuse_src` selection, which also fixes the output to more closely resemble what I see on HW (it's still not exactly right, but is much closer).

@Triticum0 if you feel up to another test pass - there's a reasonably good chance that this second change will either fix some more issues or break things that seemed to work before, totally unrelated to the alpha issue. In particular, surfaces that were untextured are the most likely to be obviously affected.

Note: This additional fix does not change the info panel in JSRF, it's still erroneously black and should be tracked as a separate issue.

**Comment by Triticum0** [comment]
I got around to retesting and there are no regressions for #242 and #598 both are still fixed. Still leave #598 open as it still plaged with another issue. The rest of the games haven't changed so are properly unrelated to this issue.

**Comment by abaire** [comment]
> I got around to retesting and there are no regressions for #242 and #598 both are still fixed. Still leave #598 open as it still plaged with another issue. The rest of the games haven't changed so are properly unrelated to this issue.

Perfect, thanks for your help!

**Comment by abaire** [comment]
> I don't fully understand the `NV_IGRAPH_XF_*` registers in `nv2a_regs.h` so I'm not sure if there's actually a HW register that should be used for the alpha value. If you can point me at docs or how to read back the register from HW I can probably figure it out. For now I just created a standalone value and piped it through to the FF shader.

Digging around docs some more, I suspect that this may be `MATERIAL_FACTOR_A` in the LTC registers listed https://envytools.readthedocs.io/en/latest/hw/graph/xf/ctx.html?highlight=LTCTX#ltctx (I'm assuming the `NV_IGRAPH_XF_LTC1_rX` and `NV_IGRAPH_XF_LTC1_LX` #defines map to `LIGHT_X_LOCAL_RANGE` and `LIGHT_X_SPOT_CUTOFF_0` resp). It looks like none of the LTC3 values are currently used so I'm not sure if it'd be desirable to put in similar handling to LTC1 or just keep a separate var until more than one thing out of LTC3 is going to be used.

**Comment by FrozenFish24** [comment]
This also fixes an issue in Ninja Gaiden Black where Ryu would incorrectly have his fiend textures applied at all times (#270).

Before:
ngb-wrong

After:
ngb

**Comment by abaire** [comment]
> This also fixes an issue in Ninja Gaiden Black where Ryu would incorrectly have his fiend textures applied at all times (#270).

Great to hear, thanks for testing @FrozenFish24 !

**Comment by abaire** [comment]
> Digging around docs some more, I suspect that this may be `MATERIAL_FACTOR_A` in the LTC registers listed https://envytools.readthedocs.io/en/latest/hw/graph/xf/ctx.html?highlight=LTCTX#ltctx (I'm assuming the `NV_IGRAPH_XF_LTC1_rX` and `NV_IGRAPH_XF_LTC1_LX` #defines map to `LIGHT_X_LOCAL_RANGE` and `LIGHT_X_SPOT_CUTOFF_0` resp). It looks like none of the LTC3 values are currently used so I'm not sure if it'd be desirable to put in similar handling to LTC1 or just keep a separate var until more than one thing out of LTC3 is going to be used.

Hey @mborgerson do you have a preference on ^^? 
I can update this to also add LTC3 (but I'd only be using the `MATERIAL_FACTOR_A` and don't know how to definitively prove that it actually is LTC3 that's being used).

**Comment by mborgerson** [comment]
> > Digging around docs some more, I suspect that this may be `MATERIAL_FACTOR_A` in the LTC registers listed https://envytools.readthedocs.io/en/latest/hw/graph/xf/ctx.html?highlight=LTCTX#ltctx (I'm assuming the `NV_IGRAPH_XF_LTC1_rX` and `NV_IGRAPH_XF_LTC1_LX` #defines map to `LIGHT_X_LOCAL_RANGE` and `LIGHT_X_SPOT_CUTOFF_0` resp). It looks like none of the LTC3 values are currently used so I'm not sure if it'd be desirable to put in similar handling to LTC1 or just keep a separate var until more than one thing out of LTC3 is going to be used.
> 
> Hey @mborgerson do you have a preference on ^^? I can update this to also add LTC3 (but I'd only be using the `MATERIAL_FACTOR_A` and don't know how to definitively prove that it actually is LTC3 that's being used).

We can keep it in a special var for now

**Comment by pigeon4195** [comment]
> Adds handling for the `SET_MATERIAL_ALPHA` command and the `diffuse_src` in the fixed function lighting model.
> 
> The test suite is here: https://github.com/abaire/nxdk_pgraph_tests/blob/main/tests/material_alpha_tests.cpp
> 
> Fixes #397 Screenshot_20211217_232350
> 
> (From @Triticum0) Fixes #242

The test suite link gives me a 404

**Comment by abaire** [comment]
I refactored the test suite sources since this PR was created. New location is https://github.com/abaire/nxdk_pgraph_tests/blob/7643b865548572c7750575d25e6c85b69283d647/src/tests/material_alpha_tests.cpp#L1

File: hw/xbox/nv2a/shaders.c [hunk 1]
```diff
@@ -488,12 +488,23 @@ GLSL_DEFINE(materialEmissionColor, GLSL_LTCTXA(NV_IGRAPH_XF_LTCTXA_CM_COL) ".xyz
 
         //FIXME: Do 2 passes if we want 2 sided-lighting?
```
Comment (MasonT8198) [inline hw/xbox/nv2a/shaders.c:489]: Would it be possible to also add this into the PR, seems easy enough

Comment (dracc) [reply hw/xbox/nv2a/shaders.c:489]: 2 sided lighting requires more NV2A methods to be implemented AFAICT. Please keep scope of PR limited. :)

Comment (MasonT8198) [reply hw/xbox/nv2a/shaders.c:489]: OIC, It looked simpler than that to me

Comment (abaire) [reply hw/xbox/nv2a/shaders.c:489]: > 2 sided lighting requires more NV2A methods to be implemented AFAICT. Please keep scope of PR limited. :)

+1, even if it was trivial it wouldn't really be related to material alpha, so it'd belong in a separate PR.

@MasonT8198 I'm curious if you know of a test case where 2-sided lighting is actually used? I've never seen it enabled in any of the nv2a logs I've examined (not that I've always paid a ton of attention). 
My copy of Forza will hopefully arrive next week and from #708 I see that @dracc came across it while log diving that game - would be a great help to add a screenshot or repro case to that bug if you know of one.

**Comment by mborgerson** [comment]
Awesome, thanks!

File: hw/xbox/nv2a/shaders.c (hunk 1 above)
Comment (MasonT8198) [reply hw/xbox/nv2a/shaders.c:489]: Forza is the only one I can think of off the top of my head, perhaps one of the games that used FFP_LIGHTSPOT would have 2 sided lighting, we can continue this discussion on the discord server

**Comment by abaire** [comment]
Sounds like this fixes #469 as well

**Comment by LazyTheHedgehog** [comment]
Hey this will sound extremely stupid, but because I know nothing about coding or whatever HOW the heck do I use this? I really wanna see the damn map in JSRF already and I genuinely got no idea how to do it

**Comment by abaire** [comment]
The easiest thing to do would be to wait until the next xemu build, which
will contain this fix.

If you don't want to wait, you will need to learn to build from source:
https://xemu.app/docs/building-from-source/ Since the fix is already merge
to master you can follow those instructions exactly and it should be
included. I'd also suggest you ask these types of questions on Discord,
GitHub is not the best place to ask for this type of support.

**Comment by ko81e24wy** [comment]
test NGB in chapter 12,fiend texture were fixed,stream effect still has issue
before
2022-02-07 02-06-27 的屏幕截图
after
2022-02-07 02-07-35 的屏幕截图

**Pull request merged**